core/src/main/scala/kafka/coordinator/group/GroupCoordinatorAdapter.scala

```diff
@@ -0,0 +1,169 @@
+import org.apache.kafka.common.protocol.Errors
+import org.apache.kafka.common.record.RecordBatch
+import org.apache.kafka.common.requests.OffsetCommitRequest
+import org.apache.kafka.common.utils.Time
+import org.apache.kafka.coordinator.group.GroupCoordinatorRequestContext
+
+import java.util
+import java.util.Optional
+import java.util.concurrent.CompletableFuture
+import scala.collection.{Map, mutable}
+import scala.jdk.CollectionConverters._
+
+/**
+ * GroupCoordinatorAdapter is a thin wrapper around kafka.coordinator.group.GroupCoordinator
+ * that exposes the new org.apache.kafka.coordinator.group.GroupCoordinator interface.
+ */
+class GroupCoordinatorAdapter(
+  private val coordinator: GroupCoordinator,
+  private val time: Time
+) extends org.apache.kafka.coordinator.group.GroupCoordinator {
+
+  override def joinGroup(
+    context: GroupCoordinatorRequestContext,
+    request: JoinGroupRequestData
+  ): CompletableFuture[JoinGroupResponseData] = {
+    val future = new CompletableFuture[JoinGroupResponseData]()
+
+    def callback(joinResult: JoinGroupResult): Unit = {
+      future.complete(new JoinGroupResponseData()
+        .setErrorCode(joinResult.error.code)
+        .setGenerationId(joinResult.generationId)
+        .setProtocolType(joinResult.protocolType.orNull)
+        .setProtocolName(joinResult.protocolName.orNull)
+        .setLeader(joinResult.leaderId)
+        .setSkipAssignment(joinResult.skipAssignment)
+        .setMemberId(joinResult.memberId)
+        .setMembers(joinResult.members.asJava)
+      )
+    }
+
+    val groupInstanceId = Option(request.groupInstanceId)
+
+    // Only return MEMBER_ID_REQUIRED error if joinGroupRequest version is >= 4
+    // and groupInstanceId is configured to unknown.
+    val requireKnownMemberId = context.apiVersion >= 4 && groupInstanceId.isEmpty
+
+    val protocols = request.protocols.valuesList.asScala.map { protocol =>
+      (protocol.name, protocol.metadata)
+    }.toList
+
+    val supportSkippingAssignment = context.apiVersion >= 9
+
+    coordinator.handleJoinGroup(
+      request.groupId,
+      request.memberId,
+      groupInstanceId,
+      requireKnownMemberId,
+      supportSkippingAssignment,
+      context.clientId,
+      context.clientAddress.toString,
+      request.rebalanceTimeoutMs,
+      request.sessionTimeoutMs,
+      request.protocolType,
+      protocols,
+      callback,
+      Option(request.reason),
+      RequestLocal(context.bufferSupplier)
+    )
+
+    future
+  }
+
+  override def commitOffsets(
+    context: GroupCoordinatorRequestContext,
+    request: OffsetCommitRequestData
+  ): CompletableFuture[OffsetCommitResponseData] = {
+    val future = new CompletableFuture[OffsetCommitResponseData]()
+
+    def callback(commitStatus: Map[TopicPartition, Errors]): Unit = {
+      val response = new OffsetCommitResponseData()
+      val byTopics = new util.HashMap[String, OffsetCommitResponseData.OffsetCommitResponseTopic]()
```

Comment:
I noticed that `byTopic` use `java.util.HasMap` while `partitions` in Line:129 use scala `mutable.HasMap` any reason why we are using two different Maps? Can't we use here `mutable.HashMap` for both?